### PR TITLE
feat: Support non-default cluster domain

### PR DESCRIPTION
**Pull request by maltesander**
# Description

Part of https://github.com/stackabletech/issues/issues/652

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [ ] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -24,12 +24,12 @@ serde = { version = "1.0", features = ["derive"] }
...
 strum = { version = "0.26", features = ["derive"] }
 tokio = { version = "1.40", features = ["full"] }
 tracing = "0.1"
 url = { version = "2.5.2" }
 xml-rs = "0.8"
 
-# [patch."https://github.com/stackabletech/operator-rs.git"]
-# stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "main" }
+[patch."https://github.com/stackabletech/operator-rs.git"]
+stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "spike/domain-in-client" }
```
Comment (NickLarsenNZ) [inline Cargo.toml:35]: Oops

```suggestion
# stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "main" }
```

File: rust/crd/src/lib.rs [hunk 2]
```diff
@@ -637,10 +641,13 @@ pub struct PodRef {
...
-    pub fn fqdn(&self) -> String {
+    pub fn fqdn(&self, cluster_info: &KubernetesClusterInfo) -> String {
         format!(
-            "{}.{}.{}.svc.cluster.local",
-            self.pod_name, self.role_group_service_name, self.namespace
+            "{}.{}.{}.svc.{}",
+            self.pod_name,
+            self.role_group_service_name,
+            self.namespace,
+            cluster_info.cluster_domain
```
Comment (NickLarsenNZ) [inline rust/crd/src/lib.rs:650]: nit: We should avoid unnamed interpolations (as they then depend on order which is fragile).

```suggestion
            "{pod_name}.{service_name}.{namespace}.svc.{cluster_domain}",
            pod_name = self.pod_name,
            service_name = self.role_group_service_name,
            namespace = self.namespace,
            cluster_domain = cluster_info.cluster_domain
```

You're welcome to leave it for now, as there are other occurrences anyway.

**Review by NickLarsenNZ — APPROVED** [review]
LGTM

File: rust/crd/src/lib.rs (hunk 2 above)
Comment (sbernauer) [reply rust/crd/src/lib.rs:650]: Was expecting this :)

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,9 @@ All notable changes to this project will be documented in this file.
 - Allow configuring proxy host behavior ([#668]).
 - Support disabling the `create-reporting-task` Job ([#690]).
 - Support podOverrides on the `create-reporting-task` Job using the field `spec.clusterConfig.createReportingTaskJob.podOverrides` ([#690]).
+- The operator can now run on Kubernetes clusters using a non-default cluster domain. It should automatically detect the
+  correct domain to use, but you can also use the env var `KUBERNETES_CLUSTER_DOMAIN` to set the domain explicitly
+  or use the helm-chart property `kubernetesClusterDomain` ([#694]).
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply CHANGELOG.md]: removed

File: deploy/helm/nifi-operator/values.yaml [hunk 4]
```diff
@@ -47,3 +47,8 @@ nodeSelector: {}
 tolerations: []
 
 affinity: {}
+
+# When running on a non-default Kubernetes cluster domain and the auto detection is not working correctly,
+# you can set your custom cluster domain here.
+# See the https://docs.stackable.tech/home/stable/guides/kubernetes-cluster-domain guide for details
+# kubernetesClusterDomain: my-cluster.local
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply deploy/helm/nifi-operator/values.yaml]: removed

**Review by sbernauer — APPROVED** [review]

**Pull request merged**